### PR TITLE
Wpf documentation for 4.8 features and bug fixes

### PR DESCRIPTION
**Pull request by miguep**
Wpf documentation for 4.8 features and bug fixes.

Accessibility features and bug fixes are bundled in one file

//cc @PriyaPurkayastha

### CODE REVIEW

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+         protected override int GetPositionInSetCore() 
+        { 
+            // Call into your own logic to provide a value for PositionInSet 
+            return CalculatePositionInSet(); 
+        } 
+    } 
+```
+3. __Automatic Values__<br/>Items in [ItemsControls](xref:System.Windows.Controls.ItemsControl) will provide a value for these properties automatically without additional action from the developer. If an ItemsControl is grouped, the collection of groups will be represented as a set and each group counted as a separate set, with each item inside that group providing it�s position inside that group as well as the size of the group. Automatic values are not affected by virtualization. Even if an item is not realized, it is still counted towards the total size of the set and affects the position in the set of it�s sibling items.
+
+##### ControllerFor property support
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
**ControllerFor property support**
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+UIAutomation�s ControllerFor property returns an array of automation elements that are manipulated by the automation element that supports this property. This property is commonly used for Auto-suggest accessibility. ControllerFor is used when an automation element affects one or more segments of the application UI or the desktop. Otherwise, it is hard to associate the impact of the control operation with UI elements. This feature adds the ability for controls to provide a value for ControllerFor property.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
UIAutomation’s ControllerFor property returns an array of automation elements that are manipulated by the automation element that supports this property. This property is commonly used for Auto-suggest accessibility. ControllerFor is used when an automation element affects one or more segments of the application UI or the desktop. Otherwise, it is hard to associate the impact of the control operation with UI elements. This feature adds the ability for controls to provide a value for the ControllerFor property.
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+A new virtual method has been added to [AutomationPeer](xref:System.Windows.Automation.Peers.AutomationPeer):
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
A new virtual method has been added to the <xref:System.Windows.Automation.Peers.AutomationPeer> class:
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+```csharp
+    virtual protected List<AutomationPeer> GetControlledPeersCore()
+```
+To provide a value for the ControllerFor property, simply override this method and return a list of [AutomationPeers](xref:System.Windows.Automation.Peers.AutomationPeer) for the controls being manipulated by this [AutomationPeer](xref:System.Windows.Automation.Peers.AutomationPeer):
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
To provide a value for the ControllerFor property, simply override this method and return a list of [AutomationPeers](xref:System.Windows.Automation.Peers.AutomationPeer) for the controls being manipulated by this <xref:System.Windows.Automation.Peers.AutomationPeer>:
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+```csharp
+
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
 ```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+            List<AutomationPeer> controlledPeers = new List<AutomationPeer>();
+            AutoSuggestTextBox owner = Owner as AutoSuggestTextBox;
+
+            controlledPeers.Add(UIElementAutomationPeer.CreatePeerForElement(owner.SuggestionListBox));
+            return controlledPeers;
+        }
+    }
+```
+
+##### Name Automation property is now correctly propagated for DatePicker and RibbonComboBox controls
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
**The Name Automation property is now correctly propagated for the DatePicker and RibbonComboBox controls**
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+Fixed an accessibility problem where [DatePicker](xref:System.Windows.Controls.DatePicker) and [RibbonComboBox](xref:System.Windows.Controls.Ribbon.RibbonComboBox) wouldn't reflect the [AutomationProperties.Name](xref:System.Windows.Automation.AutomationProperties.Name) value assigned to it when read by a screen reader.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
Starting with .NET Framework 4.8, the <xref:System.Windows.Controls.DatePicker> and <xref:System.Windows.Controls.Ribbon.RibbonComboBox> controls reflect the <xref:System.Windows.Automation.AutomationProperties.Name?displayProperty=nameWithType> value assigned to them when read by a screen reader.
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+##### Expanders in GroupItem headers are now announced by screen readers
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
**Expanders in GroupItem headers are now announced by screen readers**
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+Fixed an accessibility problem where an [Expander](xref:System.Windows.Controls.Expander) in a re-templated [GroupItem's](xref:System.Windows.Controls.GroupItem) header was not being announced by screen readers.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
Starting with .NET Framework 4.8, an <xref:System.Windows.Controls.Expander> in a re-templated <xref:System.Windows.Controls.GroupItem> header is announced by screen readers.
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+
+##### Improved focus move logic when WPF is hosted in Windows Forms
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
**Improved focus move logic when WPF is hosted in Windows Forms**
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+Fixed an accessibility problem where focus loops inside a WPF [UserControl](xref:System.Windows.Controls.UserControl) instead of breaking out of it under some hosting scenarios
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
The focus no longer loops inside a WPF <xref:System.Windows.Controls.UserControl> instead of breaking out of it under some hosting scenarios.
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+
+##### Changed padding header in GridView to not focusable
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
**Changed the padding header in GridView to not focusable**
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+Fixed an accessibility problem where the automatically generated padding header in [GridView](xref:System.Windows.Controls.GridView) was focusable, creating a bad experience for keyboard-only users.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
In the .NET Framework 4.7.2 and earlier versions, the automatically generated padding header in a <xref:System.Windows.Controls.GridView> control was focusable, creating a bad experience for keyboard-only users. Starting with .NET Framework 4.8, it is not focusable.
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+##### Fixed automation tree for plain ItemsControl
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]: **Fixed automation tree for plain ItemsControl**
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+Fixed an accessibility problem where the automation tree exposed for a plain [ItemsControls](xref:System.Windows.Controls.ItemsControl) (as opposed to a derived class like ListBox, DataGrid, etc.) was not correct.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
Fixed an accessibility problem where the automation tree exposed for a plain <xref:System.Windows.Controls.ItemsControl> (as opposed to a derived class like ListBox or DataGrid) was not correct.
```
Can this be described as something other than a fix?

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+
+##### Fixed bounding rectangles in Per-Monitor DPI.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
**Fixed bounding rectangles in Per-Monitor DPI**
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+Fixed an accessibility problem where certain bounding rectangles around text are drawn incorrectly by Accessibility tools when WPF is run in Per-Monitor Aware mode.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
In .NET Framework 4.7.2 and earlier versions, certain bounding rectangles around text are drawn incorrectly by Accessibility tools when WPF is run in Per-Monitor Aware mode. Starting with .NET Framework 4.8, they are drawn correctly.
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+</br>
+
+- [x] Quirked
+- [ ] Build-time break
+
+### Recommended Action
+__How to opt in or out of these changes__
+  
+In order for the application to benefit from these changes, it must run on the .NET Framework 4.8 or later. The application can benefit from these changes in either of the following ways:
+- It is recompiled to target the .NET Framework 4.8. These accessibility changes are enabled by default on Windows Forms applications that target the .NET Framework 4.8 or later.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
- It is recompiled to target the .NET Framework 4.8. These accessibility changes are enabled by default on WPF applications that target the .NET Framework 4.8 or later.
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+- It targets the .NET Framework 4.7.2 or earlier version and opts out of the legacy accessibility behaviors by adding the following [AppContext Switch](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element) to the `<runtime>` section of the app config file and setting it to `false`, as the following example shows.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
- It targets the .NET Framework 4.7.2 or an earlier version and opts out of the legacy accessibility behaviors by adding the following [AppContext switch](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element) to the `<runtime>` section of the app config file and setting it to `false`, as the following example shows.
```

File: Documentation/compatibility/wpf-accessibility-improvements-48.md [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+      <startup>
+        <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.7"/>
+      </startup>
+      <runtime>
+        <!-- AppContextSwitchOverrides value attribute is in the form of 'key1=true|false;key2=true|false  -->
+        <AppContextSwitchOverrides value="Switch.UseLegacyAccessibilityFeatures=false;Switch.UseLegacyAccessibilityFeatures.2=false;Switch.UseLegacyAccessibilityFeatures.3=false" />
+      </runtime>
+    </configuration>
+```    
+Note that to opt in to the accessibility features added in .NET Framework 4.8, you must also opt in to accessibility features of .NET Framework 4.7.2 and .NET Framework 4.7.1 as well.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-accessibility-improvements-48.md]:
```suggestion
Note that to opt in to the accessibility features added in .NET Framework 4.8, you must also opt in to the accessibility features of .NET Framework 4.7.2 and .NET Framework 4.7.1 as well.
```

File: Documentation/compatibility/wpf-databinding-should-use-key-not-indexer.md [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+NotPlanned
+
+### Change Description
+Fixed [Binding's](xref:System.Windows.Data.Binding) incorrect use of IList indexer when the source object declares a custom indexer with the same signature (e.g. KeyedCollection<int,TItem>).
+
+- [ ] Quirked
+- [ ] Build-time break
+
+### Recommended Action
+In order for the application to benefit from these changes, it must run on the .NET Framework 4.7.2 or later and it must opt-in to enable by setting the following [AppContext Switch](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element) to the `<runtime>` section of the app config file and setting it to `false`, as the following example shows.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-databinding-should-use-key-not-indexer.md]: Is 4.7.2 correct. If so, this should be a separate PR that merges to the master branch, not the net48 branch.

```suggestion
In order for the application to benefit from this change, it must run on the .NET Framework 4.7.2 or later, and it must opt in to the change by adding the following [AppContext switch](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element) to the `<runtime>` section of the app config file and setting it to `false`:
```

File: Documentation/compatibility/wpf-grid-row-size-allocation-with-collapsed-element.md [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+Major
+
+### Version Introduced
+4.8
+
+### Source Analyzer Status
+NotPlanned
+
+### Change Description
+Fixed a bug in the [algorithm](wpf-grid-allocation-of-space-to-star-columns.md) for allocating sizes to *-rows in a [Grid](xref:System.Windows.Controls.Grid) introduced in .NET Framework 4.7.  In some cases (e.g. Grid with Height="Auto" containing empty rows), rows were arranged at the wrong position, possibly outside the Grid altogether.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-grid-row-size-allocation-with-collapsed-element.md]:
```suggestion
Fixed a bug in the [algorithm for allocating sizes to \*-rows](wpf-grid-allocation-of-space-to-star-columns.md) in a <xref:System.Windows.Controls.Grid> introduced in .NET Framework 4.7.  In some cases, such as a Grid with `Height="Auto"` containing empty rows, rows were arranged at the wrong position, possibly outside the Grid altogether.
```

File: Documentation/compatibility/wpf-hang-listbox-duplicate-valuetypes.md [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+Major
+
+### Version Introduced
+4.8
+
+### Source Analyzer Status
+NotPlanned
+
+### Change Description
+Fixed a problem where a virtualizing [ItemsControls](xref:System.Windows.Controls.ItemsControl) can hang during scrolling when its Items collection contains duplicate value-typed objects.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-hang-listbox-duplicate-valuetypes.md]:
```suggestion
Fixed a problem where a virtualizing<xref:System.Windows.Controls.ItemsControl> can hang during scrolling when its Items collection contains duplicate value-typed objects.
```

File: Documentation/compatibility/wpf-keyboard-navigation-listbox-hyperlink.md [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+Major
+
+### Version Introduced
+4.8
+
+### Source Analyzer Status
+NotPlanned
+
+### Change Description
+Fixed incorrect result of pressing an arrow key when focus is on a hyperlink within an item that is not the [SelectedItem](xref:System.Windows.Controls.SelectedItemColleciton) of the parent [ItemsControls](xref:System.Windows.Controls.ItemsControl).
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-keyboard-navigation-listbox-hyperlink.md]:
```suggestion
Fixed incorrect result of pressing an arrow key when the focus is on a hyperlink within an item that is not the [SelectedItem](xref:System.Windows.Controls.SelectedItemCollection) of the parent <xref:System.Windows.Controls.ItemsControl>.
```

File: Documentation/compatibility/wpf-performace-improvement-grouping-automation.md [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+Major
+
+### Version Introduced
+4.8
+
+### Source Analyzer Status
+NotPlanned
+
+### Change Description
+Improved the performance of rebuilding the automation tree of an [ItemsControl](xref:System.Windows.Controls.ItemsControl) (ListBox, DataGrid, etc.) in which grouping is enabled.
```
Comment (rpetrusha) [inline Documentation/compatibility/wpf-performace-improvement-grouping-automation.md]:
```suggestion
Improved the performance of rebuilding the automation tree of an <xref:System.Windows.Controls.ItemsControl>, such as a ListBox or DataGrid, in which grouping is enabled. 
```

File: Documentation/compatibility/wpf-HwndHost-child-hwnd-correct-size-during-DPI-changes.md [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+NotPlanned
+
+### Change Description
+Prior to .NET Framework 4.8, when WPF was run in Per-Monitor Aware mode, controls hosted within [HwndHost](xref:System.Windows.Interop.HwndHost) were not sized correctly after DPI changes (for e.g., when moving applications from one monitor to another). This fix ensures that controls hosted so are sized appropriately. 
+
+- [x] Quirked
+- [ ] Build-time break
+
+### Recommended Action
+In order for the application to benefit from these changes, it must run on the .NET Framework 4.7.2 or later and it must opt-in to enable by setting the following [AppContext Switch](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element) to the `<runtime>` section of the app config file and setting it to `false`, as the following example shows.
```
[the remark this answers is not in the document]
Comment (miguep) [reply Documentation/compatibility/wpf-HwndHost-child-hwnd-correct-size-during-DPI-changes.md]: This is a bug fix, made for 4.8, but also requested for backport to 4.7.2

File: Documentation/compatibility/wpf-HwndHost-child-hwnd-correct-size-during-DPI-changes.md [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+
+```xml
+    <?xml version="1.0" encoding="utf-8"?>
+    <configuration>
+      <startup>
+        <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.7"/>
```
[the remark this answers is not in the document]
Comment (miguep) [reply Documentation/compatibility/wpf-HwndHost-child-hwnd-correct-size-during-DPI-changes.md]: no, this illustrates the scenario where your application targets an older framework version yet still receives the 4.8 fixes when setting the AppContext switch

File: Documentation/compatibility/wpf-SelectionTextBrush-property-for-non-adorner-selection.md [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+Major
+
+### Version Introduced
+4.8
+
+### Source Analyzer Status
+NotPlanned
+
+### Change Description
+In WPF applications using non-adorner based text selection for [TextBox](xref:System.Windows.Controls.TextBox) and [PasswordBox](xref:System.Windows.Controls.PasswordBox)  via [AppContext Switch](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element) "Switch.System.Windows.Controls.Text.UseAdornerForTextboxSelectionRendering=false", developers may now set the SelectionTextBrush property in order to alter the rendering of the selected text.  By default, this color changes with [SystemColors.HighlightTextBrushKey](xref:System.Windows.SystemColors.HighlightTextBrushKey).  If non-adorner based text selection is not enabled, this property does nothing.
```
[the remark this answers is not in the document]
Comment (miguep) [reply Documentation/compatibility/wpf-SelectionTextBrush-property-for-non-adorner-selection.md]: The behavior didn't change, there is a new property that enables changing the selected text's brush, but it only applies to non-adorner based text selection, a feature introduced in 4.7.2

**Comment by miguep** [comment]
@rpetrusha I've addressed the comments, please let me know if the modifications are ok

**Comment by rpetrusha** [comment]
Thanks for making the additional changes, @miguep. I'll merge your PR now.
